File: php/enlighten/bin/SelfTest.php
<?php

require_once(__DIR__ . '/../vendor/autoload.php');

use GuzzleHttp\Client;
use GuzzleHttp\HandlerStack;
use OpenAPI\Client\Api\DefaultApi;
use OpenAPI\Client\Configuration;
use Spatie\GuzzleRateLimiterMiddleware\RateLimiterMiddleware;

parse_str(implode('&', array_slice($argv, 1)), $_GET);
$uid = $_GET['--user_id'];
$cfg = Configuration::getDefaultConfiguration()
    ->setApiKey('key', $_GET['--api_key'])
    ->setDebug(true)
    ->setDebugFile('php://stderr');
$stack = HandlerStack::create();
$stack->push(RateLimiterMiddleware::perMinute(10));
$cli = new Client(['handler' => $stack]);
$api = new DefaultApi($cli, $cfg);

$startDate = (new DateTime)->sub(new DateInterval('P1M'))->format('Y-m-d');
$endDate = (new DateTime)->sub(new DateInterval('P3W'))->format('Y-m-d');
$startAt = (new DateTime)->sub(new DateInterval('P1W'))->getTimestamp();
$endAt = (new DateTime)->sub(new DateInterval('P1D'))->getTimestamp();

$next = null;
while (true) {
    $res = $api->systems($uid, $next);
    foreach ($res->getSystems() as $sys) {
        $res2 = $api->systems($uid, null, 1,
            $sys->getSystemId(), null,
            $sys->getSystemName(), null,
            $sys->getStatus(), null,
            $sys->getReference(), null,
            null, null,
            $sys->getConnectionType());
        if (sizeof($res2->getSystems()) !== 1) {
            die('systems: expected one result');
        }
        $api->invertersSummaryByEnvoyOrSite($uid, $sys->getSystemId());
        $api->energyLifetime($uid, $sys->getSystemId());
        $api->energyLifetime($uid, $sys->getSystemId(), $startDate, $endDate, 'all');
        foreach ($api->envoys($uid, $sys->getSystemId())->getEnvoys() as $env) {
            $api->searchSystemId($uid, $env->getSerialNumber());
        }
        $api->inventory($uid, $sys->getSystemId());
        $api->monthlyProduction($uid, $sys->getSystemId(), $startDate);
        $api->productionMeterReadings($uid, $sys->getSystemId());
        $api->productionMeterReadings($uid, $sys->getSystemId(), $endAt);
        $api->rgmStats($uid, $sys->getSystemId());
        $api->rgmStats($uid, $sys->getSystemId(), $startAt, $endAt);
        $api->stats($uid, $sys->getSystemId());
        $api->stats($uid, $sys->getSystemId(), $startAt, $endAt);
        $api->consumptionLifetime($uid, $sys->getSystemId());
        $api->consumptionLifetime($uid, $sys->getSystemId(), $startDate, $endDate);
        $api->consumptionStats($uid, $sys->getSystemId());
        $api->consumptionStats($uid, $sys->getSystemId(), $startAt, $endAt);
        $api->summary($uid, $sys->getSystemId());
        $api->summary($uid, $sys->getSystemId(), $startDate);
    }
    $next = $res->getNext();
    if ($next === null) {
        break;
    }
}

File: python/enlighten/selftest.py
import argparse
import sys
from datetime import date, datetime

import ratelimit
from dateutil.relativedelta import relativedelta

import enlighten
from enlighten.api import default_api

parser = argparse.ArgumentParser()
parser.add_argument('--api_key')
parser.add_argument('--user_id')
args = parser.parse_args()
uid = args.user_id
cfg = enlighten.Configuration(api_key={'ApiKey': args.api_key})
cfg.debug = True

with enlighten.ApiClient(cfg) as cli:
    cli.call_api = ratelimit.sleep_and_retry(ratelimit.limits(calls=10, period=60)(cli.call_api))
    api = default_api.DefaultApi(cli)

    start_date = date.today() - relativedelta(months=1)
    end_date = date.today() - relativedelta(weeks=3)
    start_at = int((datetime.now() - relativedelta(weeks=1)).timestamp())
    end_at = int((datetime.now() - relativedelta(days=1)).timestamp())

    next = ''
    while True:
        res = api.systems(uid)
        for s in res.systems:
            kwargs = {'system_id': s.system_id, 'status': s.status, 'connection_type': s.connection_type}
            if hasattr(s, 'system_name'):
                kwargs['system_name'] = s.system_name
            if hasattr(s, 'reference'):
                kwargs['reference'] = s.reference
            if len(api.systems(uid, limit=1, **kwargs).systems) != 1:
                sys.exit('systems: expected one result')

            api.inverters_summary_by_envoy_or_site(uid, s.system_id)
            api.energy_lifetime(uid, s.system_id)
            api.energy_lifetime(uid, s.system_id, start_date=start_date, end_date=end_date, production='all')

            for env in api.envoys(uid, s.system_id).envoys:
                api.search_system_id(uid, env.serial_number)

            api.inventory(uid, s.system_id)
            api.monthly_production(uid, s.system_id, start_date)
            api.production_meter_readings(uid, s.system_id)
            api.production_meter_readings(uid, s.system_id, end_at=end_at)
            api.rgm_stats(uid, s.system_id)
            api.rgm_stats(uid, s.system_id, start_at=start_at, end_at=end_at)
            api.stats(uid, s.system_id)
            api.stats(uid, s.system_id, start_at=start_at, end_at=end_at)
            api.consumption_lifetime(uid, s.system_id)
            api.consumption_lifetime(uid, s.system_id, start_date=start_date, end_date=end_date)
            api.consumption_stats(uid, s.system_id)
            api.consumption_stats(uid, s.system_id, start_at=start_at, end_at=end_at)
            api.summary(uid, s.system_id)
            api.summary(uid, s.system_id, summary_date=start_date)
        next = getattr(res, 'next', '')
        if next == '':
            break
